test/Microsoft.ML.Tests/Transformers/TextFeaturizerTests.cs

```diff
@@ -51,6 +52,139 @@ public void TextFeaturizerWithPredefinedStopWordRemoverTest()
+
+            var prediction = engine.Predict(data[0]);
+            Assert.Equal("this is some text in english", string.Join(" ", prediction.OutputTokens));
+            Assert.Equal(1.0f, prediction.Features[0]);
+            Assert.Equal(1.0f, prediction.Features[1]);
+            Assert.Equal(1.0f, prediction.Features[3]);
+            Assert.Equal(1.0f, prediction.Features[4]);
+            Assert.Equal(1.0f, prediction.Features[5]);
+            Assert.Equal(0.0f, prediction.Features[6]);
+            Assert.Equal(0.0f, prediction.Features[7]);
+
+            prediction = engine.Predict(data[1]);
+            Assert.Equal("this is another example", string.Join(" ", prediction.OutputTokens));
+            Assert.Equal(1.0f, prediction.Features[0]);
+            Assert.Equal(1.0f, prediction.Features[1]);
+            Assert.Equal(0.0f, prediction.Features[3]);
+            Assert.Equal(0.0f, prediction.Features[4]);
+            Assert.Equal(0.0f, prediction.Features[5]);
+            Assert.Equal(1.0f, prediction.Features[6]);
+            Assert.Equal(1.0f, prediction.Features[7]);
+        }
+
+        [Fact]
+        public void TextFeaturizerWithCharFeatureExtractorTest()
+        {
+            var data = new[] { new TestClass() { A = "abc efg", OutputTokens=null},
+                               new TestClass() { A = "xyz", OutputTokens=null } };
+            var dataView = ML.Data.LoadFromEnumerable(data);
+
+            var options = new TextFeaturizingEstimator.Options()
+            {
+                WordFeatureExtractor = null,
+                CharFeatureExtractor = new WordBagEstimator.Options() { NgramLength = 1 },
+                Norm = TextFeaturizingEstimator.NormFunction.None,
+                OutputTokensColumnName = "OutputTokens"
+            };
+            var pipeline = ML.Transforms.Text.FeaturizeText("Features", options, "A");
+            var model = pipeline.Fit(dataView);
+            var engine = model.CreatePredictionEngine<TestClass, TestClass>(ML);
+
+            var prediction = engine.Predict(data[0]);
+            Assert.Equal("abc efg", string.Join(" ", prediction.OutputTokens));
+            Assert.Equal(1.0f, prediction.Features[0]);
+            Assert.Equal(1.0f, prediction.Features[1]);
+            Assert.Equal(1.0f, prediction.Features[3]);
+            Assert.Equal(1.0f, prediction.Features[4]);
+            Assert.Equal(1.0f, prediction.Features[5]);
+            Assert.Equal(1.0f, prediction.Features[6]);
+            Assert.Equal(1.0f, prediction.Features[7]);
+            Assert.Equal(1.0f, prediction.Features[8]);
+            Assert.Equal(0.0f, prediction.Features[9]);
+            Assert.Equal(0.0f, prediction.Features[10]);
+            Assert.Equal(0.0f, prediction.Features[11]);
+
+            prediction = engine.Predict(data[1]);
+            Assert.Equal("xyz", string.Join(" ", prediction.OutputTokens));
+            Assert.Equal(1.0f, prediction.Features[0]);
```

Comment:
>Assert.Equal(1.0f, prediction.Features[0]); [](start = 12, length = 43)

So i expect ngrams to be `a`,`b`,`c`,`e`,`f`,`g`,`x`,`y`,`z`,` ` and end of string and empty string, not sure honestly,  I guess. total 12 ngrams.
feature 0 and feature 8 is this end of string and empty string, right?

Reply:
Index 0 is start marker and index 8 is end marker. Then there are total 10 characters including space.

---
In reply to: [267486364](https://github.com/dotnet/machinelearning/pull/3036#discussion_r267486364) [](ancestors = 267486364)